src/adc.rs

```diff
@@ -336,27 +336,31 @@ macro_rules! adc_hal {
 
                 /// sets up adc in one shot mode for a single channel
                 pub fn setup_oneshot(&mut self) {
-                    // stop and clear overrun events
-                    self.rb.cr.modify(|_, w| w.adstp().set_bit());
-                    self.rb.isr.modify(|_, w| w.ovr().clear_bit());
+                    self.rb.cr.modify(|_, w| w.adstp().stop());
+                    self.rb.isr.modify(|_, w| w.ovr().clear());
 
                     self.rb.cfgr.modify(|_, w| w
-                        .cont().clear_bit()
-                        .ovrmod().clear_bit()
+                        .cont().single()
+                        .ovrmod().preserve()
                     );
 
-                    self.rb.sqr1.modify(|_, w| w.l().bits(0));
+                    self.set_sequence_len(1);
 
                     self.operation_mode = Some(OperationMode::OneShot);
                 }
 
+                fn set_sequence_len(&mut self, len: u8) {
+                    debug_assert!(len <= 16);
```

Comment:
Since you use `len - 1` below, you need to check that `len - 1 < 16`. Or add an additional check that `len > 0`.